crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -237,9 +233,11 @@ pub struct ReadState<T> {
 unsafe impl<T: Component> FetchState for ReadState<T> {
     fn init(world: &mut World) -> Self {
         let component_info = world.components.get_or_insert_info::<T>();
+        if T::Storage::STORAGE_TYPE == StorageType::SparseSet {
```

Comment:
Adding a registration here (and not elsewhere, like WriteState) seems a bit odd. Maybe we need some sort of `world.init_component::<T>() -> ComponentId` function that does a get_or_insert operation _and_ sets up storages if necessary?  Maybe nothing should call `components.get_or_insert_info::<T>()` directly?